### PR TITLE
[AMDGPU] Form V_MAD_U64_U32 from mul24

### PR DESCRIPTION
**Pull request by Pierre-vh**
Fixes SWDEV-421067

### CODE REVIEW

**Comment by jayfoad** [comment]
> CGP can transform a fine mul+add into a (mul24/mulhi24)+add, so add a pattern for that.

Typo "fine"? Not sure what you meant.

This would depend on the relative rate of mul_u24 vs mad_u64. On older ASICs, mad_u64 is "quarter rate" so two fast mul_u24 instructions should be faster. I see that gfx90a uses SIDPFullSpeedModel so mad_u64 is as fast as mul_u24.

In any case would it be better to teach CGP not to do the harmful transformation int he first place, rather than work around it in isel?

**Comment by Pierre-vh** [comment]
> > CGP can transform a fine mul+add into a (mul24/mulhi24)+add, so add a pattern for that.
> 
> Typo "fine"? Not sure what you meant.
> 
> This would depend on the relative rate of mul_u24 vs mad_u64. On older ASICs, mad_u64 is "quarter rate" so two fast mul_u24 instructions should be faster. I see that gfx90a uses SIDPFullSpeedModel so mad_u64 is as fast as mul_u24.
> 
> In any case would it be better to teach CGP not to do the harmful transformation int he first place, rather than work around it in isel?

By fine I meant "fine to transform into v_mad" but I didn't know about FullSpeed/QuarterSpeed - I edited it.
Should I add a predicate on this to only do the transform on FullSpeedModels then?

About doing this in CGP, I asked @arsenm earlier and he suggested to fix it in ISel rather than teach CGP. I tend to agree - CGP doesn't always have the full picture/full knowledge of what the DAG can do, and it may be non-obvious to fix in CGP. Having a new pattern is simpler and more stable, IMO

**Comment by jayfoad** [comment]
> About doing this in CGP, I asked @arsenm earlier and he suggested to fix it in ISel rather than teach CGP. I tend to agree - CGP doesn't always have the full picture/full knowledge of what the DAG can do, and it may be non-obvious to fix in CGP. Having a new pattern is simpler and more stable, IMO

You might both be right (I'm not super familiar with CGP) but I don't really understand why. On a "full rate doubles" machine (i.e. SIDPFullSpeedModel or SIDPGFX940FullSpeedModel), 24-bit multiplies are no faster than 32-bit, so why would CGP need to do anything at all?

**Comment by arsenm** [comment]
> You might both be right (I'm not super familiar with CGP) but I don't really understand why. On a "full rate doubles" machine (i.e. SIDPFullSpeedModel or SIDPGFX940FullSpeedModel), 24-bit multiplies are no faster than 32-bit, so why would CGP need to do anything at all?

I believe they also use less power. Plus better to not depend on the specific speed model, the more common looking ISA the better

**Comment by jayfoad** [comment]
For background: my understanding is that the only reason CGP needs to help select 24-bit multiplies is that you can't do it during instruction selection because there is no way for a target to simplify a generic opcode like ISD::MUL based on demanded bits information. See https://discourse.llvm.org/t/selectiondag-target-specific-simplification-of-generic-nodes-using-demanded-bits/56747/1

With the current patch I still don't like the fact that CGP will split a 64-bit multiply and expect isel to combine it back into a single instruction later. How about this for a cleaner design:
- Change llvm.amdgcn.mul.i24 and llvm.amdgcn.mul.u24 to work for any scalar integer type, not just i32. So they are like a regular multiply but only look at the low 24 bits of their inputs. (Or the result is undefined if the inputs are not signed/unsigned 24-bit values - I'm not sure if that is a more useful definition?)
- Then CGP can convert mul to llvm.amdgcn.mul.i24/u24 where possible with no need to split up 64-bit multiplies.
- Isel patterns with suitable predicates based on the speed model can choose to implement a 64-bit i24/u24 multiply as either a pair of 24-bit multiplies, or a single 64-bit mad.

**Comment by Pierre-vh** [comment]
> For background: my understanding is that the only reason CGP needs to help select 24-bit multiplies is that you can't do it during instruction selection because there is no way for a target to simplify a generic opcode like ISD::MUL based on demanded bits information. See https://discourse.llvm.org/t/selectiondag-target-specific-simplification-of-generic-nodes-using-demanded-bits/56747/1
> 
> With the current patch I still don't like the fact that CGP will split a 64-bit multiply and expect isel to combine it back into a single instruction later. How about this for a cleaner design:
> 
> * Change llvm.amdgcn.mul.i24 and llvm.amdgcn.mul.u24 to work for any scalar integer type, not just i32. So they are like a regular multiply but only look at the low 24 bits of their inputs. (Or the result is undefined if the inputs are not signed/unsigned 24-bit values - I'm not sure if that is a more useful definition?)
> * Then CGP can convert mul to llvm.amdgcn.mul.i24/u24 where possible with no need to split up 64-bit multiplies.
> * Isel patterns with suitable predicates based on the speed model can choose to implement a 64-bit i24/u24 multiply as either a pair of 24-bit multiplies, or a single 64-bit mad.

Should the mulhi_24 intrinsics be gone or stay in that case?

If I understand correctly we'd only end up removing the second pattern, the first one would stay right?

Also I still don't understand the deal with the speed model. Should it affect ISel patterns really?
Or do we just do this:
 - V_MAD pattern can select a mul24 
 - mul24 not selected by any other pattern are lowered to a pair of mul24/mul24hi

**Comment by jayfoad** [comment]
> Also I still don't understand the deal with the speed model. Should it affect ISel patterns really?

Yes, I think so. The job of isel (including pattern-based isel) is to pick the "best" instructions according to some cost model, either performance or code size, so certainly changes in the cost model should affect which isel patterns are used.

**Comment by Pierre-vh** [comment]
@jayfoad I did the refactor you suggested, seems to be looking better indeed.

However I still haven't done the SpeedModel part because:
- (Apologies but) I'm still confused. Do we need it on those new patterns (we only form the V_MAD_I64_I32 if on FullSpeed) or in #72983 to emit a i64 mul instead of i24 muls on non-FullSpeed models?
- We don't have patterns like that currently so I'm wondering if there is a good reason behind it? (cc @arsenm)

**Comment by jayfoad** [comment]
> However I still haven't done the SpeedModel part because:
> 
> * (Apologies but) I'm still confused. Do we need it on those new patterns (we only form the V_MAD_I64_I32 if on FullSpeed) or in [[AMDGPU] Don't create mulhi_24 in CGP #72983](https://github.com/llvm/llvm-project/pull/72983) to emit a i64 mul instead of i24 muls on non-FullSpeed models?

The new patterns should probably have a FullRate64Ops predicate, and _if_ you want to use V_MAD_I64_I32 for a plain 32 * 32 -> 64 bit multiply (not a multiply-add), as suggested inline, then the old pattern could also have a NotFullRate64Ops predicate. It all depends on exactly which pattern you want to be used for each speed model, and making sure there is always at least one pattern that will match each DAG node so you don't get selection failures.

Also, as an alternative to putting predicates on patterns, you could try fiddling with AddedComplexity (since that's the closest thing we have to a cost model for pattern-based selection). But it probably doesn't make much difference in practice.

> * We don't have patterns like that currently so I'm wondering if there is a good reason behind it? (cc @arsenm)

I guess there aren't many cases where we want to select different instructions to do the same thing for different speed models.

**Comment by arsenm** [comment]
> For background: my understanding is that the only reason CGP needs to help select 24-bit multiplies is that you can't do it during instruction selection because there is no way for a target to simplify a generic opcode like ISD::MUL based on demanded 

I think I originally moved this because computeKnownBits in the IR is significantly better than on the DAG. In particular because of the cross block copy issue

File: llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll [hunk 1]
```diff
@@ -6928,7 +6962,119 @@ entry:
...
+; GFX900-SDAG-NEXT:    v_addc_co_u32_e32 v1, vcc, v4, v3, vcc
+; GFX900-SDAG-NEXT:    s_setpc_b64 s[30:31]
+;
+; GFX9-GISEL-LABEL: mul_u24_add64:
+; GFX9-GISEL:       ; %bb.0:
+; GFX9-GISEL-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-GISEL-NEXT:    v_mul_hi_u32_u24_e32 v4, v0, v1
+; GFX9-GISEL-NEXT:    v_mul_u32_u24_e32 v0, v0, v1
+; GFX9-GISEL-NEXT:    v_add_co_u32_e32 v0, vcc, v0, v2
+; GFX9-GISEL-NEXT:    v_addc_co_u32_e32 v1, vcc, v4, v3, vcc
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll:6999]: I don't follow how this is better? This is 16 cycles? vs. 16 cycles for the v_mad_u64_u32? We could also consider optsize

File: llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll [hunk 2]
```diff
@@ -6928,7 +6962,119 @@ entry:
...
+; GFX10-LABEL: mul_u24_zext_add64:
+; GFX10:       ; %bb.0:
+; GFX10-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX10-NEXT:    v_mul_u32_u24_e32 v0, v0, v1
+; GFX10-NEXT:    v_add_co_u32 v0, vcc_lo, v0, v2
+; GFX10-NEXT:    v_add_co_ci_u32_e32 v1, vcc_lo, 0, v3, vcc_lo
+; GFX10-NEXT:    s_setpc_b64 s[30:31]
+  %mul = call i32 @llvm.amdgcn.mul.u24(i32 %x, i32 %y)
+  %mul.zext = zext i32 %mul to i64
+  %add = add nuw nsw i64 %mul.zext, %z
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll:7069]: Drop the flags, pretty sure you don't need them

File: llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll (hunk 1 above)
Comment (Pierre-vh) [reply llvm/test/CodeGen/AMDGPU/integer-mad-patterns.ll:6999]: @jayfoad using SpeedModel was your idea, what do you think?

**Comment by Pierre-vh** [comment]
Small ping

**Comment by Pierre-vh** [comment]
@jayfoad, can you answer the speed model questions? I don't know enough about it to answer those. Thanks

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 3]
```diff
@@ -676,6 +676,16 @@ multiclass IMAD32_Pats <VOP3_Pseudo inst> {
...
         (EXTRACT_SUBREG (inst $src0, $src1, (i64 (as_i64imm $src2)), 0 /* clamp */), sub0)
         >;
+
+  // Handle cases where amdgpu-codegenprepare-mul24 made a mul24 instead of a normal mul.
+  def : GCNPat <
+      (i64 (add (i64 (AMDGPUmul_u24 i32:$src0, i32:$src1)), i64:$src2)),
+      (inst $src0, $src1, $src2, 0 /* clamp */)
+      >;
+  def : GCNPat <
+      (i64 (add (i64 (zext (i32 (AMDGPUmul_u24 i32:$src0, i32:$src1)))), i64:$src2)),
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:689]: Should we do an extract_subreg and fill the upper 32 bits with zero in the result instruction then?

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 4]
```diff
@@ -676,6 +676,16 @@ multiclass IMAD32_Pats <VOP3_Pseudo inst> {
         (ThreeOpFragSDAG<mul, add> i32:$src0, i32:$src1, (i32 imm:$src2)),
         (EXTRACT_SUBREG (inst $src0, $src1, (i64 (as_i64imm $src2)), 0 /* clamp */), sub0)
         >;
+
+  // Handle cases where amdgpu-codegenprepare-mul24 made a mul24 instead of a normal mul.
+  def : GCNPat <
+      (i64 (add (i64 (AMDGPUmul_u24 i32:$src0, i32:$src1)), i64:$src2)),
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:685]: Didn't see the signed version in the ticket this is intended to fix so I'd prefer to leave it out of this diff

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 5]
```diff
@@ -676,6 +676,16 @@ multiclass IMAD32_Pats <VOP3_Pseudo inst> {
         (ThreeOpFragSDAG<mul, add> i32:$src0, i32:$src1, (i32 imm:$src2)),
         (EXTRACT_SUBREG (inst $src0, $src1, (i64 (as_i64imm $src2)), 0 /* clamp */), sub0)
         >;
+
+  // Handle cases where amdgpu-codegenprepare-mul24 made a mul24 instead of a normal mul.
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: No strong preference, if it works and it's faster then sure let's also add that

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td (hunk 3 above)
[the remark this answers is not in the document]
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:689]: Let's step back a bit: what are you trying to achieve with this pattern? Why did you add it and what ISA do you want it to generate?

[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:689]: Initially, I was trying to get V_MAD out of:
```
          t328: i32 = MUL_U24 # D:1 t430, Constant:i32<22853>
          t103: i32 = llvm.amdgcn.mulhi.u24 # D:1 TargetConstant:i64<2489>, t430, Constant:i32<22853>
        t419: v2i32 = BUILD_VECTOR # D:1 t328, t103
      t420: i64 = bitcast # D:1 t419
          t428: i32 = truncate # D:1 t255
        t426: v2i32 = BUILD_VECTOR # D:1 t428, Constant:i32<0>
      t427: i64 = bitcast # D:1 t426
    t108: i64 = add # D:1 t420, t427
```

Now we no longer have the `BUILD_VECTOR` due to the CGP changes.
I think the i32 mul24 pattern you're asking about isn't needed - I just thought I'd also make it work for i32. I'll retest without it to see if it matters at all.

[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: I removed that pattern, it's not needed. I re-verified the ticket just to be safe.
Let's only try to get the pattern that matter in this patch.

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 6]
```diff
@@ -678,9 +678,26 @@ multiclass IMAD32_Pats <VOP3_Pseudo inst> {
         >;
 }
 
+// Handle cases where amdgpu-codegenprepare-mul24 made a mul24 instead of a normal mul.
+// We need to separate this because otherwise OtherPredicates would be overriden.
+multiclass IMAD32_Mul24_Pats <VOP3_Pseudo inst> {
+  def : GCNPat <
+      (i64 (add (i64 (AMDGPUmul_u24 i32:$src0, i32:$src1)), i64:$src2)),
+      (inst $src0, $src1, $src2, 0 /* clamp */)
+      >;
```
[the remark this answers is not in the document]
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: Agreed, if the multiply is followed by a full 64 bit add then we should always use mad_u64 regardless of speed model.

**Review by jayfoad — APPROVED** [review]

**Pull request merged**